consensus/polybft/state.go

```diff
@@ -256,6 +318,79 @@ func (s *State) list() ([]*StateSyncEvent, error) {
 	return events, nil
 }
 
+// insertStateSyncEvent inserts a new state sync event to state event bucket in db
+func (s *State) insertExitEvent(event *ExitEvent) error {
+	return s.db.Update(func(tx *bolt.Tx) error {
+		raw, err := json.Marshal(event)
+		if err != nil {
+			return err
+		}
+
+		bucket := tx.Bucket(exitEventsBucket)
+
+		return bucket.Put(bytes.Join([][]byte{itob(event.Epoch), itob(event.ID), itob(event.BlockNumber)}, nil), raw)
+	})
+}
+
+// getExitEvent returns exit event with given id, which happened in given epoch and given block number
+func (s *State) getExitEvent(exitEventID, epoch, checkpointBlockNumber uint64) (*ExitEvent, error) {
+	var exitEvent *ExitEvent
+
+	err := s.db.View(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(exitEventsBucket)
+
+		key := bytes.Join([][]byte{itob(epoch), itob(exitEventID), itob(checkpointBlockNumber)}, nil)
+		v := bucket.Get(key)
+		if v == nil {
+			return &ExitEventNotFoundError{exitEventID, checkpointBlockNumber, epoch}
+		}
+
+		return json.Unmarshal(v, &exitEvent)
+	})
+
+	return exitEvent, err
+}
+
+// getExitEventsByEpoch returns all exit events that happened in the given epoch
+func (s *State) getExitEventsByEpoch(epoch uint64) ([]*ExitEvent, error) {
+	return s.getExitEvents(epoch, func(exitEvent *ExitEvent) bool {
+		return exitEvent.Epoch == epoch
+	})
+}
+
+// getExitEventsForProof returns all exit events that happened in and prior to the given checkpoint block number
+// with respect to the epoch in which block is added
+func (s *State) getExitEventsForProof(epoch, checkpointBlock uint64) ([]*ExitEvent, error) {
+	return s.getExitEvents(epoch, func(exitEvent *ExitEvent) bool {
+		return exitEvent.Epoch == epoch && exitEvent.BlockNumber <= checkpointBlock
+	})
+}
+
+// getExitEvents returns exit events for given epoch and provided filter
+func (s *State) getExitEvents(epoch uint64, filter func(exitEvent *ExitEvent) bool) ([]*ExitEvent, error) {
+	var events []*ExitEvent
+
+	err := s.db.View(func(tx *bolt.Tx) error {
+		c := tx.Bucket(exitEventsBucket).Cursor()
+		prefix := itob(epoch)
+
+		for k, v := c.Seek(prefix); bytes.HasPrefix(k, prefix); k, v = c.Next() {
+			var event *ExitEvent
+			if err := json.Unmarshal(v, &event); err != nil {
+				return err
+			}
+
+			if filter(event) {
```

Comment:
if we want to make a proof up to specific block(N) - we'll have to read all epoch events. After block N we can stop reading new exit events.

Reply:
That is what we are doing. Take a look at the function `getExitEventsForProof`.

Reply:
Yeah, I see your point now, so I was taking a look at this. It is not easy to do. Unfortunately, boltdb only supports seek by prefix :disappointed:, and since we need all the events that happened in that epoch, until given block, there is no optimization good enough. Even if I put block number at the first place in key, we would still need to get events that have block equal or lower than given block plus their epochs have to match. To me, what we have now is the most optimal we can get from boltdb in this case.